File: scripts/rust_controller/src/shared.rs
//! Shared arm state. `Shared::angles` is the single set of motor angles every
//! input (window sliders, joystick, IK, the web page, RIFT, remote clients)
//! writes to and every output (serial, the sliders, the web page) reads from.
//!
//! The sync rule: an input whose value changed since last frame writes it
//! (last writer wins); inputs that didn't change leave what another input
//! wrote alone and instead follow it. All maps are keyed by motor number.

use std::collections::BTreeMap;
use std::sync::{Arc, Mutex, MutexGuard};

#[derive(Default)]
pub struct Shared {
    pub angles: BTreeMap<i32, i32>,
    /// Client mode only: local changes not yet pushed to the hub.
    pub pending: BTreeMap<i32, i32>,
}

pub type SharedState = Arc<Mutex<Shared>>;

pub fn lock(state: &SharedState) -> MutexGuard<'_, Shared> {
    // A panic elsewhere shouldn't take the whole arm down with a poisoned lock.
    state.lock().unwrap_or_else(|e| e.into_inner())
}

/// Writes every entry of `local` that differs from `prev_local` into `shared`,
/// remembers `local` for next frame, and returns what it changed.
pub fn merge_local(
    shared: &mut BTreeMap<i32, i32>,
    prev_local: &mut BTreeMap<i32, i32>,
    local: &BTreeMap<i32, i32>,
) -> BTreeMap<i32, i32> {
    let mut changed = BTreeMap::new();
    for (&n, &angle) in local {
        if prev_local.get(&n) != Some(&angle) {
            shared.insert(n, angle);
            changed.insert(n, angle);
        }
    }
    *prev_local = local.clone();
    changed
}

#[cfg(test)]
mod tests {
    use super::*;

    fn map(pairs: &[(i32, i32)]) -> BTreeMap<i32, i32> {
        pairs.iter().copied().collect()
    }

    #[test]
    fn first_frame_writes_everything() {
        let mut shared = map(&[(1, 135), (2, 50)]);
        let mut prev = BTreeMap::new();
        let changed = merge_local(&mut shared, &mut prev, &map(&[(1, 100), (2, 50)]));
        assert_eq!(shared, map(&[(1, 100), (2, 50)]));
        assert_eq!(changed, map(&[(1, 100), (2, 50)]));
    }

    #[test]
    fn untouched_inputs_leave_remote_changes_alone() {
        let mut shared = map(&[(1, 135), (2, 50)]);
        let mut prev = map(&[(1, 135), (2, 50)]);
        shared.insert(1, 140); // the phone moves motor 1
        let changed = merge_local(&mut shared, &mut prev, &map(&[(1, 135), (2, 50)]));
        assert_eq!(shared[&1], 140);
        assert!(changed.is_empty());
    }

    #[test]
    fn a_moved_input_wins() {
        let mut shared = map(&[(1, 140), (2, 50)]);
        let mut prev = map(&[(1, 135), (2, 50)]);
        let changed = merge_local(&mut shared, &mut prev, &map(&[(1, 135), (2, 80)]));
        assert_eq!(shared, map(&[(1, 140), (2, 80)])); // motor 2 moved locally, motor 1 stays remote
        assert_eq!(changed, map(&[(2, 80)]));
    }
}

File: scripts/go_controller/shared.go
package main

// Shared arm state. Shared.angles is the single set of motor angles every
// input (window sliders, joystick, IK, the web page, RIFT, remote clients)
// writes to and every output (serial, the sliders, the web page) reads from.
//
// The sync rule: an input whose value changed since last frame writes it
// (last writer wins); inputs that didn't change leave what another input wrote
// alone and instead follow it. All maps are keyed by motor number.

import "sync"

type Shared struct {
	mu      sync.Mutex
	angles  map[int]int
	pending map[int]int // client mode only: local changes not yet pushed to the hub
}

func NewShared(angles map[int]int) *Shared {
	cp := make(map[int]int, len(angles))
	for n, a := range angles {
		cp[n] = a
	}
	return &Shared{angles: cp, pending: map[int]int{}}
}

func (s *Shared) Snapshot() map[int]int {
	s.mu.Lock()
	defer s.mu.Unlock()
	cp := make(map[int]int, len(s.angles))
	for n, a := range s.angles {
		cp[n] = a
	}
	return cp
}

func (s *Shared) Set(motor, angle int) {
	s.mu.Lock()
	s.angles[motor] = angle
	s.mu.Unlock()
}

// Step adds delta to motor's angle, kept within [lo, hi], and returns the new angle.
func (s *Shared) Step(motor, delta, lo, hi int) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	a := s.angles[motor] + delta
	if a < lo {
		a = lo
	} else if a > hi {
		a = hi
	}
	s.angles[motor] = a
	return a
}

func (s *Shared) Get(motor int) (int, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	a, ok := s.angles[motor]
	return a, ok
}

// MergeLocal writes every entry of local that differs from *prev into the
// shared angles, remembers local for next frame, and returns what it changed.
func (s *Shared) MergeLocal(prev *map[int]int, local map[int]int) map[int]int {
	changed := map[int]int{}
	s.mu.Lock()
	for n, a := range local {
		if old, ok := (*prev)[n]; !ok || old != a {
			s.angles[n] = a
			changed[n] = a
		}
	}
	s.mu.Unlock()
	next := make(map[int]int, len(local))
	for n, a := range local {
		next[n] = a
	}
	*prev = next
	return changed
}

// Queue records local changes that a client must push to its hub.
func (s *Shared) Queue(changed map[int]int) {
	s.mu.Lock()
	for n, a := range changed {
		s.pending[n] = a
	}
	s.mu.Unlock()
}
